src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -629,6 +630,14 @@ public static string Join(string? separator, IEnumerable<string?> values)
             {
                 throw new ArgumentNullException(nameof(values));
             }
+            if (values is List<string?> valuesIList)
+            {
+                return Join(separator, CollectionsMarshal.AsSpan(valuesIList), 0, valuesIList.Count);
```

Comment:
This could corrupt memory in the face of a race condition. If the list is being added to concurrently, the count could end up being larger than the span; the helper then uses unsafe code to walk the list,  up to the count. The code needs to use the same span for the count.

Reply:
Actually,  looking at the Assam implementation,  it's not even safe for this use.

Reply:
Oh. I misread the helper. It's not using unsafe code to walk the list,  just the separator. Ok.

Reply:
Fixed by passing count inside the helper method